cert/aziot-certd/src/lib.rs

```diff
@@ -106,15 +123,61 @@ impl Api {
             return Err(Error::Unauthorized(user, id));
         }
 
-        let x509 = create_cert(
-            &mut *this,
-            &id,
-            &csr,
-            issuer
-                .as_ref()
-                .map(|(issuer_cert, issuer_private_key)| (&**issuer_cert, issuer_private_key)),
-        )
-        .await?;
+        let req = X509Req::from_pem(&csr)
+            .map_err(|err| Error::invalid_parameter("csr", err))?;
+        let pubkey = req.public_key()
+            .map_err(|err| Error::invalid_parameter("csr", err))?;
+
+        if !req.verify(&pubkey)
+            .map_err(|err| Error::invalid_parameter("csr", err))?
+        {
+            return Err(Error::invalid_parameter(
+                "csr",
+                "CSR failed to be verified with its public key".to_owned()
+            ))
+        }
+
+        let issuer = issuer
+            .map(|(id, handle)| -> Result<_, Error> {
+                let pem = get_cert_inner(
+                        &this.homedir_path,
+                        &this.preloaded_certs,
+                        &id
+                    )
+                    .map_err(|err| Error::Internal(InternalError::CreateCert(err.into())))?
+                    .ok_or_else(|| Error::invalid_parameter("issuer.certId", "not found"))?;
+                let stack = X509::stack_from_pem(&pem)
+                    .map_err(|err| Error::Internal(InternalError::CreateCert(err.into())))
+                    .and_then(|stack| {
+                        stack.get(0)
+                            .ok_or_else(||
+                                Error::invalid_parameter("issuer.certId", "invalid issuer")
+                            )?;
+                        Ok(stack)
+                    })?;
+
+                let handle = CString::new(handle.0)
+                    .map_err(|err| Error::invalid_parameter("issuer.privateKeyHandle", err))?;
+
+                Ok((stack, handle))
+            })
+            .transpose()?;
+
+        let x509 = create_cert_inner(
+                &mut *this,
+                &id,
+                (&req, &pubkey),
+                issuer.as_ref().map(|(id, handle)| (&**id, &**handle))
```

Comment:
```suggestion
                issuer.as_ref().map(|(x509, pk)| (&**x509, &**pk))
```

Reply:
Fixed in 56ac22b39cba8804ead30ca1d3031387e08366b9.